test/Test/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -287,6 +297,36 @@ prop_roundtrip csize ps numEntries =
 
     showBS = unlines . showBytes . LBS.unpack
 
+prop_total_deserialisation :: [Word32] -> Property
+prop_total_deserialisation word32s =
+    let !(ByteArray ba) = byteArrayFromList word32s
+    in case fromSBS (SBS.SBS ba) of
+      Left err -> label err $ property True
+      Right (numEntries, ci) -> label "parsed successfully" $ property $
+        -- Just forcing the index is not enough. The underlying vectors might
+        -- point to outside of the byte array, so we check they are valid.
+        (numEntries, ci) `deepseq`
+             vec32IsValid (ciRangeFinder ci)
+          && vec32IsValid (ciPrimary ci)
+          && bitVecIsValid (ciClashes ci)
+          && bitVecIsValid (ciLargerThanPage ci)
+  where
+    vec32IsValid (VU.V_Word32 (VP.Vector off len ba)) =
+      off >= 0 && len >= 0 && mul4 (off + len) <= sizeofByteArray ba
+    bitVecIsValid (BV.BitVec off len ba) =
+      off >= 0 && len >= 0 && ceilDiv8 (off + len) <= sizeofByteArray ba
+
+prop_total_deserialisation_whitebox :: RFPrecision -> Small Word16 -> Small Word16 -> [Word32] -> Property
+prop_total_deserialisation_whitebox (RFPrecision rfprec) numEntries numPages word32s =
+    prop_total_deserialisation $
+         [1]  -- version
+      <> word32s  -- primary array, range finder, clash bits, LTP bits, clash map
+      <> [0 | even (length word32s)]  -- padding
+      <> [ fromIntegral rfprec, 0
+         , fromIntegral numPages, 0
+         , fromIntegral numEntries, 0
+         ]
```

Comment:
This deserialises succesfully only with a very small probability. However, since we're doing roundtrip serialisation/deserialisation tests elsewhere, that's probably not a bad thing. Since the serialisation format for the compact index is relatively complex (compared to the bloom filter, for example), it would be quite hard anyway to generate inputs that make the parse succeed

Reply:
Yes, I tried for a little bit, but also realised that the error cases are the more important ones here.